File: 프로그래머스/0/181881. 조건에 맞게 수열 변환하기 2/조건에 맞게 수열 변환하기 2.js
function solution(arr) {
    let count = 0;
    let changed = true;

    while (changed) {
        changed = false;
        let newArr = arr.slice(); // 현재 상태 복사

        for (let i = 0; i < arr.length; i++) {
            let num = arr[i];

            if (num >= 50 && num % 2 === 0) {
                newArr[i] = num / 2;
            } else if (num < 50 && num % 2 === 1) {
                newArr[i] = num * 2 + 1;
            }
        }

        // 변화가 없으면 종료
        for (let i = 0; i < arr.length; i++) {
            if (arr[i] !== newArr[i]) {
                changed = true;
                break;
            }
        }

        if (changed) {
            arr = newArr;
            count++;
        }
    }

    return count;
}

File: 프로그래머스/0/181881. 조건에 맞게 수열 변환하기 2/조건에 맞게 수열 변환하기 2.kt
class Solution {
    fun solution(arr: IntArray): Int {
        var count = 0
        var changed = true
        var current = arr.copyOf()

        while (changed) {
            changed = false
            val next = current.copyOf()

            for (i in current.indices) {
                val value = current[i]
                if (value >= 50 && value % 2 == 0) {
                    next[i] = value / 2
                } else if (value < 50 && value % 2 == 1) {
                    next[i] = value * 2 + 1
                }
            }

            for (i in current.indices) {
                if (current[i] != next[i]) {
                    changed = true
                    break
                }
            }

            if (changed) {
                current = next
                count++
            }
        }

        return count
    }
}
